### DISCUSSION TITLE
IF_DATA parsing

### DISCUSSION THREAD

**Discussion by Ganganthi** [Luncher91/A2LParser#21] [General]
Hi, I need to parse an A2L file to a JSON file, but some fields such as IF_DATAs are not fully parsed. I'd like to implement it by myself. Which file uses the ANTLR generated files, creates the objects and prints the json? Is it available in the repo?

**Comment by Luncher91** [comment]
Hi,

I like the idea to parse the content of a IF_DATA block. I was not able to find a complete specification for that language. If you would like to do it and maybe consider to make a pull request - that would be awesome.

I give you a quick overview:
- the maven antlr4 plugin generates multiple files like a2lLexer.java, a2lParser.java and a2lParserBaseVisitor.java
- A2LVisitor extends a2lParserBaseVisitor and overwrites most of its methods
- The A2LVisitor generates a Java object structure starting with the Asap2File and all its sub structures
- From this object structure the json is serialized using the com.fasterxml.jackson.databind.ObjectMapper (see Asap2File.toJson methods)

The command line interface is provided by Asap2Parser

I hope I have been able to give you a kick start here ;-)

Kind regards,
Andreas

**Reply by Ganganthi** [reply]
Hi,

Thank you for the quick reply. I'm pretty new to both java and ANTLR, so I might have some beginner questions. I am thinking of doing the following:

- Updating the if_data_block on the grammar files (.g4) and generating new Parser and Lexer files with antlr4
- Altering the ifData.java file to work with the new structure (and possibly creating new files for different if_data blocks)

Is there anything else that I have to change?

Where are all the files generated by antlr4? Are they mixed inside the src/main/java/net/alenzen/a2l/ folder?

I have also seen that A2LVisitor.java is where the parsed information is stored inside the objects, but I can't find where are those methods called. Is it something that antlr does that I'm unaware of? (This question might be pointless if this file was generated by antlr4)

Kind regards,
Thiago.

**Reply by Luncher91** [reply]
Hi Thiago,

 I am no ANTLR expert either so I cannot tell you what the best way is to tackle that but I can tell you how I would approach it. I would separate the parsing of the IF_DATA content completely from the parsing of the rest of the A2l. You can have a look at what I did with the formula parsing recently. The API for that is not yet finished but I think it was a good idea to separate the parsing.

The generated code will be generated into target/generate-sources/antlr4/ and then contained in the package net.alenzen.a2l.antlr
This directory will then also be used as a source fille directory.
A good starting point is to run `mvn clean build`
This will fetch all the dependencies and builds the source code after generating the sources from the g4 files. You may want to run it in a docker container if you like.

To understand the parsing workflow please have a look at `Asap2Parser.parse()`
There are in general 3 steps:
- Lexing (cut the input into tokens)
- Parsing (match the tokens and generate a parse tree)
- Visiting (do whatever you need to do with the parsed data)

The A2LVisitor receives the ParseTree from the Parser. The base class implements most of the visiting process and I overwrote the methods I needed to gather the information to generate the object structure.

You see the same steps for the Formula parsing in FormulaValidator.validate()

I hope I was able to give you an overview about how the parsing works for this project.
Disclaimer: This is pretty mich my first rather good approach to ANTLR and might not the best way to use ANTLR.

Further information:
ANTLR 4 Maven Plugin: https://www.antlr.org/api/maven-plugin/latest/index.html
If you are new to java you might also be new to maven: https://maven.apache.org/what-is-maven.html

Kind regards,
Andreas
